src/storage/models.rs

```diff
@@ -0,0 +1,261 @@
+use std::iter;
+
+use chrono::{DateTime, Utc};
+use rand::Rng;
+
+const DEFAULT_SLUG_LENGTH: usize = 8;
+
+/// A code snippet
+#[derive(Debug, Default, Eq, PartialEq)]
+pub struct Snippet {
+    /// Slug that uniquely identifies the snippet
+    pub id: String,
+    /// Snippet title. None means that the title was not specified at snippet
+    /// creation time
+    pub title: Option<String>,
+    /// Snippet syntax. None means that the syntax was not specified at snippet
+    /// creation time
+    pub syntax: Option<String>,
+    /// List of revisions. The snippet content is the content of the most recent
+    /// revision. Revisions are sorted by the version number in the
+    /// ascending order.
+    pub changesets: Vec<Changeset>,
+    /// List of tags attached to the snippet
+    pub tags: Vec<String>,
+    /// Timestamp of when the snippet was created
+    pub created_at: Option<DateTime<Utc>>,
+    /// Timestamp of when the snippet was last modified. Defaults to creation
+    /// time
+    pub updated_at: Option<DateTime<Utc>>,
+}
+
+impl Snippet {
+    /// Create a new Snippet.
+    pub fn new(
+        title: Option<String>,
+        syntax: Option<String>,
+        changesets: Vec<Changeset>,
+        tags: Vec<String>,
+    ) -> Self {
+        let mut snippet = Snippet::default();
+        snippet.id = Snippet::random_id(DEFAULT_SLUG_LENGTH);
+        snippet.title = title;
+        snippet.syntax = syntax;
+        snippet.changesets = changesets;
+        snippet.tags = tags;
+
+        snippet
+    }
+
+    /// Generate a random unique snippet identifier (slug).
+    pub fn random_id(length: usize) -> String {
+        let mut rng = rand::thread_rng();
+        iter::repeat_with(|| rng.sample(rand::distributions::Alphanumeric))
+            .take(length)
+            .collect()
+    }
+}
+
+/// A particular snippet revision
+#[derive(Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Changeset {
+    /// Changeset index. Version numbers start from 0 and are incremented by 1
+    pub version: usize,
+    /// Changeset content
+    pub content: String,
+    /// Timestamp of when the changeset was created
+    pub created_at: Option<DateTime<Utc>>,
+    /// Timestamp of when the changeset was last modified. Defaults to creation
+    /// time
+    pub updated_at: Option<DateTime<Utc>>,
+}
+
+impl Changeset {
+    /// Create a new Changeset.
+    pub fn new(version: usize, content: String) -> Self {
+        let mut changeset = Changeset::default();
+        changeset.version = version;
+        changeset.content = content;
+
+        changeset
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```

Comment:
TBH, I'm not a fan of that Rust idea of writing test in the same module where the code lies. :-/

Reply:
We totally could put this into a separate module and include it here (so that the test module would have access to the private functions/structs/fields of this one), but that would not be idiomatic, AFAIK :(